src/app/modules/form/admin-form/admin-form.service.ts

```diff
@@ -204,55 +205,81 @@ export const getMockSpcpLocals = (
  * @param newOwnerEmail the email of the new owner to transfer to
  *
  * @return ok(updated form) if transfer is successful
+ * @return err(MissingUserError) if either the current user or user to transfer to cannot be found in the database
  * @return err(TransferOwnershipError) if new owner cannot be found in the database or new owner email is same as current owner
  * @return err(DatabaseError) if any database errors like missing admin of current owner occurs
  */
 export const transferFormOwnership = (
   currentForm: IPopulatedForm,
   newOwnerEmail: string,
-): ResultAsync<IPopulatedForm, TransferOwnershipError | DatabaseError> => {
+): ResultAsync<
+  IPopulatedForm,
+  MissingUserError | TransferOwnershipError | DatabaseError
+> => {
   const logMeta = {
     action: 'transferFormOwnership',
     newOwnerEmail,
   }
 
-  // Step 1: Transfer form ownership.
-  return ResultAsync.fromPromise(
-    currentForm.transferOwner(newOwnerEmail),
-    (error) => {
-      logger.error({
-        message: 'Error occurred whilst transferring form ownership',
-        meta: logMeta,
-        error,
-      })
-
-      // Special case, when document instance method already returns
-      // ApplicationError, directly return the error.
-      if (
-        error instanceof TransferOwnershipError ||
-        error instanceof DatabaseError
-      ) {
-        return error
-      }
+  return (
+    // Step 1: Retrieve current owner of form to transfer.
+    findAdminById(currentForm.admin._id)
+      .andThen((currentOwner) =>
+        // Step 2: Retrieve user document for new owner.
+        findAdminByEmail(newOwnerEmail)
+          .mapErr((error) => {
+            // Override MissingUserError with more specific message if new owner
+            // cannot be found.
+            if (error instanceof MissingUserError) {
+              error.message = `${newOwnerEmail} must have logged in once before being added as Owner`
```

Comment:
providing the correct error message to the frontend should be the responsibility of the controller, not the service

Reply:
The problem with this is that MissingUserError can be thrown in various stages, and different stages require a different error message :/. The controller will not know which stage the error was thrown from. Was it from failing to retrieve the current logged in user or the new to-be owner?

Reply:
just split into separate steps instead of chaining? errors should be treated as immutable, repurposing them like this is a bad code smell. if this pattern appeared everywhere we could even lose information in stack traces.

Reply:
Since this is not an async function, chaining must occur. What I've done instead is return a new `TransferFormOwnership` error instead of rewriting the message... would that be better? See 9c11b20